### PR TITLE
mgr/rook: update `ceph orch apply nfs` 

### PR DESCRIPTION
**Pull request by josephsawaya**
This PR updates the `ceph orch apply nfs` command in the Rook orchestrator to keep up with the changes made to the NFS module by preventing the creation of NFS daemons not using the '.nfs' RADOS pool and creating the pool if it doesn't exist when a user tries to create an NFS daemon.

Depends on [rook #8501](https://github.com/rook/rook/pull/8501)

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/rook/rook_cluster.py [hunk 1]
```diff
@@ -607,12 +609,21 @@ def _update_zone(new: cos.CephObjectStore) -> cos.CephObjectStore:
...
             _update_zone, _create_zone)
 
-    def apply_nfsgw(self, spec: NFSServiceSpec) -> str:
+    def apply_nfsgw(self, spec: NFSServiceSpec, mgr: 'RookOrchestrator') -> str:
         # TODO use spec.placement
         # TODO warn if spec.extended has entries we don't kow how
         #      to action.
         # TODO Number of pods should be based on the list of hosts in the
         #      PlacementSpec.
+        service_id = spec.service_id if spec.service_id is not None else ""
```
Comment (varshar16) [inline src/pybind/mgr/rook/rook_cluster.py:618]: `service_id` should never be empty string. Add a check for both None and empty string

**Review by varshar16 — CHANGES_REQUESTED** [review]
Please address the following comments

File: src/pybind/mgr/rook/rook_cluster.py [hunk 2]
```diff
@@ -6,7 +6,9 @@
 
 This module is runnable outside of ceph-mgr, useful for testing.
 """
+from nfs.module import Module
```
Comment (varshar16) [inline src/pybind/mgr/rook/rook_cluster.py:9]: Please move this import below `from nfs.cluster import create_ganesha_pool`

File: src/pybind/mgr/rook/rook_cluster.py [hunk 3]
```diff
@@ -6,7 +6,9 @@
...
 import datetime
+from nfs.export import NFSRados
```
Comment (varshar16) [inline src/pybind/mgr/rook/rook_cluster.py:11]: Please move this import below `from nfs.cluster import create_ganesha_pool`

File: src/pybind/mgr/rook/rook_cluster.py [hunk 4]
```diff
@@ -924,12 +927,19 @@ def _update_zone(new: cos.CephObjectStore) -> cos.CephObjectStore:
...
         # TODO use spec.placement
         # TODO warn if spec.extended has entries we don't kow how
         #      to action.
         # TODO Number of pods should be based on the list of hosts in the
         #      PlacementSpec.
+        if spec.service_id == None:
+            raise RuntimeError("service id in NFS service spec cannot be None")
+        assert spec.service_id # for mypy typing
+        if spec.service_id == "":
+            raise RuntimeError("service id in NFS service spec cannot be an empty string")
```
Comment (varshar16) [inline src/pybind/mgr/rook/rook_cluster.py:940]:
```suggestion
        assert spec.service_id, "service id in NFS service spec cannot be an empty string or None " # for mypy typing
```

**Review by varshar16 — CHANGES_REQUESTED** [review]
Apart from the comment below, it would be great if you can add qa tests.

File: src/pybind/mgr/rook/rook_cluster.py [hunk 5]
```diff
@@ -25,6 +27,7 @@
 from ceph.deployment.service_spec import ServiceSpec, NFSServiceSpec, RGWSpec
 from ceph.utils import datetime_now
 from ceph.deployment.drive_selection.matchers import SizeMatcher
+from nfs.cluster import create_ganesha_pool
```
Comment (varshar16) [inline src/pybind/mgr/rook/rook_cluster.py:30]: According to [PEP 8 style guide](https://www.python.org/dev/peps/pep-0008/#imports), imports should be grouped in the following order:
Standard library imports.
Related third party imports.
Local application/library specific imports

```suggestion
from nfs.cluster import create_ganesha_pool
from nfs.module import Module
from nfs.export import NFSRados
```

**Review by varshar16 — CHANGES_REQUESTED** [review]
Please move the imports. Otherwise looks good.

**Review by varshar16 — APPROVED** [review]
Please run the qa tests after the depending PR is merged. Otherwise looks good.

File: src/pybind/mgr/rook/module.py [hunk 6]
```diff
@@ -431,6 +433,9 @@ def remove_service(self, service_name: str) -> str:
             return self.rook_cluster.rm_service('cephobjectstores', service_name)
         elif service_type == 'nfs':
             return self.rook_cluster.rm_service('cephnfses', service_name)
+        elif service_type == 'ingress':
```
Comment (jmolmo) [inline src/pybind/mgr/rook/module.py:436]: This is not harmful,, but I think that nobody is going to create an ingress service in a rook cluster, so difficult to reach this point. In any case, i think the operation that must be blocked is the creation of an orchestrator ingress service within the rook orchestrator

File: src/pybind/mgr/rook/rook_cluster.py [hunk 7]
```diff
@@ -924,12 +927,15 @@ def _update_zone(new: cos.CephObjectStore) -> cos.CephObjectStore:
...
             _update_zone, _create_zone)
 
-    def apply_nfsgw(self, spec: NFSServiceSpec) -> str:
+    def apply_nfsgw(self, spec: NFSServiceSpec, mgr: 'RookOrchestrator') -> str:
         # TODO use spec.placement
         # TODO warn if spec.extended has entries we don't kow how
         #      to action.
         # TODO Number of pods should be based on the list of hosts in the
         #      PlacementSpec.
+        assert spec.service_id, "service id in NFS service spec cannot be an empty string or None " # for mypy typing
```
Comment (jmolmo) [inline src/pybind/mgr/rook/rook_cluster.py:936]: No error management... we have commented about this problem in other PRs. Basically if we do not include a basic error management in this methods, we are reporting managing programming and unexpected errors  in the same way that errors related to the operative.
At least, we should avoid to dump the stack trace in the console ( is good to do that in the log... but in the console causes panic in final users) 

Example:

```
# ceph orch apply nfs jmo_nfs

Error EINVAL: Traceback (most recent call last):
  File "/usr/share/ceph/mgr/mgr_module.py", line 1564, in _handle_command
    return self.handle_command(inbuf, cmd)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 167, in handle_command
    return dispatch[cmd['prefix']].call(self, cmd, inbuf)
  File "/usr/share/ceph/mgr/mgr_module.py", line 415, in call
    return self.func(mgr, **kwargs)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 107, in <lambda>
    wrapper_copy = lambda *l_args, **l_kwargs: wrapper(*l_args, **l_kwargs)  # noqa: E731
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 96, in wrapper
    return func(*args, **kwargs)
  File "/usr/share/ceph/mgr/orchestrator/module.py", line 1155, in _apply_nfs
    return self._apply_misc([spec], dry_run, format, no_overwrite)
  File "/usr/share/ceph/mgr/orchestrator/module.py", line 1062, in _apply_misc
    raise_if_exception(completion)
  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 224, in raise_if_exception
    raise e
kubernetes.client.rest.ApiException: (422)
Reason: Unprocessable Entity
HTTP response headers: HTTPHeaderDict({'Audit-Id': '611aedf2-314e-4f9b-a828-1c2c5a292027', 'Cache-Control': 'no-cache, private', 'Content-Type': 'application/json', 'X-Kubernetes-Pf-Flowschema-Uid': '21f0f9d3-30f5-4bef-a93b-e25595220817', 'X-Kubernetes-Pf-Prioritylevel-Uid': 'ce436ed7-8716-4a58-8b63-07fd783fdd5c', 'Date': 'Tue, 14 Sep 2021 11:10:23 GMT', 'Content-Length': '908'})
HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"CephNFS.ceph.rook.io \"jmo_nfs\" is invalid: metadata.name: Invalid value: \"jmo_nfs\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')","reason":"Invalid","details":{"name":"jmo_nfs","group":"ceph.rook.io","kind":"CephNFS","causes":[{"reason":"FieldValueInvalid","message":"Invalid value: \"jmo_nfs\": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')","field":"metadata.name"}]},"code":422}
```

File: src/pybind/mgr/rook/module.py [hunk 8]
```diff
@@ -431,6 +433,9 @@ def remove_service(self, service_name: str) -> str:
             return self.rook_cluster.rm_service('cephobjectstores', service_name)
         elif service_type == 'nfs':
```
Comment (jmolmo) [inline src/pybind/mgr/rook/module.py:434]: Ceph auth keys for the nfs service are not deleted. This is an operation to do in **ALL** the orchestrators

```
# ceph auth ls

client.nfs-ganesha.jmonfs.a
	key: AQChg0Bh5V7ZHRAAFMydhUKjSPKvZRxV7qTBsg==
	caps: [mon] allow r
	caps: [osd] allow rw pool=.nfs namespace=jmonfs

```

**Review by jmolmo — CHANGES_REQUESTED** [review]
Apart of the issues pointed in the comments, the list of the service is not properly managed:

```
[root@rook-ceph-tools-78cdfd976c-qc6fm /]# ceph orch ls
NAME          PORTS  RUNNING  REFRESHED  AGE  PLACEMENT  
crash                    3/3  0s ago     38m  *          
mgr                      1/1  0s ago     38m  count:1    
mon                      3/3  0s ago     39m  count:3    
nfs.test2nfs             0/1  0s ago     -    count:1  <--------------  Number of nfs pods not updated properly
```

**Review by jmolmo — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
jenkins test sign

**Pull request merged**